services/messageService.js

```diff
@@ -0,0 +1,51 @@
+import { AsyncStorage } from 'react-native';
+
+class MessageService {
+  constructor(newMessagesCallback) {
+    this.messages = {};
+    this.newMessagesCallback = newMessagesCallback;
+  }
+
+  static async cacheMessages(contact, messages) {
+    try {
+      const localMessages = await AsyncStorage.getItem('messages');
+
+      let localMessageDict = {};
+      if (localMessages) {
+        localMessageDict = JSON.parse(localMessages);
+      }
+
+      localMessageDict[contact] = messages;
+      await AsyncStorage.setItem('messages', JSON.stringify(localMessageDict));
+    } catch (e) {
+      // TODO: Add sentry logging
+    }
+  }
+
+  async updateLocalMessageStateSingle(contact, message) {
+    const messages = {
+      items: [message]
+    };
+
+    this.updateLocalMessageState(contact, messages);
+  }
+
+  async updateLocalMessageState(contact, rawMessages) {
+    const messages = MessageService.formatRawMessages(rawMessages);
+    this.newMessagesCallback(messages);
+    MessageService.cacheMessages(contact, messages);
+  }
+
+  static formatRawMessages(rawMessages) {
+    return rawMessages.items.reverse().map(message => ({
```

Comment:
Is this related to that gifted chat requiring the messages in reverse order? Is there anyway to store the messages to avoid this reverse?

Reply:
Yeah the ultimate problem is that gifted chat needs the messages in one order, but the twilio API returns them in a different order. 

When we switch to redux, this will be mitigated since we will have a reducer that will have to manage slotting messages in by timestamp. 

I also opened a bug to investigate if there's a way to force the Twilio API to return the messages in a different order.